File: integration_tests/dart_client_tests_real/test/lists_test.dart
import 'package:test/test.dart';
import 'package:graphlink_client_integration_tests/generated/client/graph_link_client.dart';
import 'package:graphlink_client_integration_tests/generated/enums/user_status.dart';
import 'real_server_adapter.dart';

void main() {
  late GraphLinkClient client;
  setUp(() => client = newClient());

  group('non-nullable list of objects', () {
    test('listUsers returns a correctly typed list with 2 users', () async {
      final res = await client.queries.listUsers(limit: 10);
      expect(res.listUsers.length, equals(2));
      expect(res.listUsers[0].id, equals('user-1'));
      expect(res.listUsers[1].id, equals('user-2'));
    });

    test('each item in the list is fully deserialized', () async {
      final res = await client.queries.listUsers(limit: 10);
      final user = res.listUsers.first;
      expect(user.name, equals('Alice Smith'));
      expect(user.address.city, equals('Springfield'));
    });
  });

  group('list of scalars on a type', () {
    test('User.tags list deserializes correctly for Alice', () async {
      final res = await client.queries.getUser(id: 'user-1', limit: 10);
      expect(res.getUser.tags, equals(['admin', 'beta']));
    });

    test('empty tags list is empty for Bob', () async {
      final res = await client.queries.getUser(id: 'user-2', limit: 10);
      expect(res.getUser.tags, isEmpty);
    });

    test('nullable scores list is null for Bob', () async {
      final res = await client.queries.getUser(id: 'user-2', limit: 10);
      expect(res.getUser.scores, isNull);
    });

    test('nullable scores list has values for Alice', () async {
      final res = await client.queries.getUser(id: 'user-1', limit: 10);
      expect(res.getUser.scores, equals([10, 20, 30]));
    });
  });

  group('list of objects with nullable fields', () {
    test('getTags returns tags with nullable color', () async {
      final res = await client.queries.getTags();
      expect(res.getTags.length, equals(2));
      expect(res.getTags[0].label, equals('dart'));
      expect(res.getTags[0].color, equals('#0175C2'));
      expect(res.getTags[1].color, isNull);
    });
  });

  group('list query with enum argument', () {
    test('listUsersByStatus(ACTIVE) returns only active users', () async {
      final res = await client.queries
          .listUsersByStatus(status: UserStatus.ACTIVE, limit: 10);
      expect(res.listUsersByStatus, isNotEmpty);
      expect(res.listUsersByStatus.first.status, equals(UserStatus.ACTIVE));
    });
  });
}

File: integration_tests/typescript_client_tests_real/test/lists.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { GraphLinkClient } from '../lib/generated/client/graph-link-client.js';
import { UserStatus } from '../lib/generated/enums/user-status.js';
import { newClient } from './real-server-adapter.ts';

let client: GraphLinkClient;
beforeEach(() => { client = newClient(); });

describe('non-nullable list of objects', () => {
  it('listUsers returns 2 users', async () => {
    const res = await client.queries.listUsers();
    expect(res.listUsers).toHaveLength(2);
  });

  it('listUsers items have correct ids', async () => {
    const res = await client.queries.listUsers();
    expect(res.listUsers[0].id).toBe('user-1');
    expect(res.listUsers[1].id).toBe('user-2');
  });

  it('each item is fully deserialized', async () => {
    const res = await client.queries.listUsers();
    expect(res.listUsers[0].name).toBe('Alice Smith');
    expect(res.listUsers[0].address.city).toBe('Springfield');
  });
});

describe('list of scalars on a type', () => {
  it('User.tags list deserializes correctly for Alice', async () => {
    const res = await client.queries.getUser({ id: 'user-1' });
    expect(res.getUser.tags).toEqual(['admin', 'beta']);
  });

  it('empty tags list for Bob', async () => {
    const res = await client.queries.getUser({ id: 'user-2' });
    expect(res.getUser.tags).toHaveLength(0);
  });

  it('nullable scores is null for Bob', async () => {
    const res = await client.queries.getUser({ id: 'user-2' });
    expect(res.getUser.scores).toBeNull();
  });

  it('nullable scores has values for Alice', async () => {
    const res = await client.queries.getUser({ id: 'user-1' });
    expect(res.getUser.scores).toEqual([10, 20, 30]);
  });
});

describe('list of objects with nullable fields', () => {
  it('getTags returns 2 tags with correct fields', async () => {
    const res = await client.queries.getTags();
    expect(res.getTags).toHaveLength(2);
    expect(res.getTags[0].label).toBe('dart');
    expect(res.getTags[0].color).toBe('#0175C2');
    expect(res.getTags[1].color).toBeNull();
  });
});

describe('list query with enum argument', () => {
  it('listUsersByStatus(ACTIVE) returns only active users', async () => {
    const res = await client.queries.listUsersByStatus({ status: UserStatus.Active });
    expect(res.listUsersByStatus.every(u => u.status === UserStatus.Active)).toBe(true);
  });
});
